File: AdventOfCode/2023/Day4/partb.py
f = open("file.txt", "r")
match = []
count = 0
newMatch = []
finalList=[]
for line in f.readlines():
    cardNumber = line.split(":")[0]

    flag=False
    total = 1
    # print(cardNumber)
    firstHalf = line.split(":")[1].split("|")[0].split(" ")
    secondHalf = line.split(":")[1].split("|")[1].split(" ")

    #    from intenet
    firstHalf = [s.strip() for s in firstHalf if s.strip() != ""]
    secondHalf = [s.strip() for s in secondHalf if s.strip() != ""]


    for i in firstHalf:
        for k in secondHalf:
            if i == k:
               count+=1
                                 
    match.append(count)            
    count=0
               
print(match)

total_copies = 0
new_match = match.copy()

while new_match:
    total_copies += sum(new_match)
    for i in range(1, len(match)):
        if i < len(new_match):
           new_match.append(match[i])
print(total_copies)

File: AdventOfCode/2023/Day4/partb.js
const fs = require('fs');

let data = fs.readFileSync('file.txt', 'utf-8');
let lines = data.split('\n');
let match = [];
let newMatch = [];
let finalList = [];

for (let line of lines) {
    let splitLine = line.split(":");
    let firstHalf = splitLine[1].split("|")[0].split(" ").filter(s => s.trim() !== "");
    let secondHalf = splitLine[1].split("|")[1].split(" ").filter(s => s.trim() !== "");

    let count = firstHalf.filter(value => secondHalf.includes(value)).length;
    match.push(count);
}

for (let i = 0; i < match.length; i++) {
    let value = match[i];
    finalList.push(i + 1);
    for (let j = i + 1; j < i + 1 + value; j++) {
        if (j < match.length) {
            newMatch.push(j + 1);
            finalList.push(j + 1);
        }
    }
}

while (newMatch.length > 0) {
    let tempMatch = [];
    for (let i = 0; i < newMatch.length; i++) {
        let value = newMatch[i];
        for (let j = i + 1; j < i + 1 + value; j++) {
            if (j < match.length) {
                tempMatch.push(j + 1);
                finalList.push(j + 1);
            }
        }
    }
    newMatch = tempMatch;
}

console.log(finalList);